File: tiki_sdk_flutter_plugin/ios/Classes/TikiSdkFlutterChannel.swift
import Flutter
import Promises

public class TikiSdkFlutterChannel: FlutterMethodChannel, FlutterPlugin {

    public static let channelId = "tiki_sdk_flutter"

    public static var channel: FlutterMethodChannel?

    public var plugin: TikiSdkFlutterPlugin? = nil


    public lazy var flutterEngine = FlutterEngine(name: "tiki_sdk_flutter_engine")
    public var methodChannel: FlutterMethodChannel? = nil

    public init(apiKey: String? =  nil, origin: String? = nil) {
        super.init()
        if(methodChannel == nil){
            setupChannel(apiKey: apiKey ?? "", origin: origin ?? "")
        }
    }

    public func buildSdk(apiKey: String, origin: String) {
        plugin = TikiSdkFlutterPlugin(methodChannel: methodChannel!)
        methodChannel!.invokeMethod(
            "build", arguments: [
                "apiKey" : apiKey,
                "origin" : origin
            ]
        )
    }

    @objc(handleMethodCall:result:) public func handle(_ call: FlutterMethodCall, result: @escaping FlutterResult) {
        guard let requestId = (call.arguments as! Dictionary<String, String>)["requestId"] else {
            result(FlutterError.init(code: "-1", message: "missing requestId argument", details: call.arguments))
            return
        }
        let response = (call.arguments as? Dictionary<String, String>)?["response"]
        switch (call.method) {
            case "success" :
                plugin!.promises[requestId]!.fulfill(response)
            break
            case "error" :
                plugin!.promises[requestId]!.reject(TikiSdkError(message: response))
            break
            default : result(FlutterError(code: "-1", message: "Uninplemented", details: call.arguments))
        }
    }

    @objc(registerWithRegistrar:) public static func register(with registrar: FlutterPluginRegistrar) {
        channel = FlutterMethodChannel(name: channelId, binaryMessenger: registrar.messenger())
        registrar.addMethodCallDelegate(TikiSdkFlutterChannel(), channel: channel!)
    }


    public func setupChannel(apiKey: String, origin: String) {
        if (methodChannel == nil) {
            flutterEngine.run()
            methodChannel = FlutterMethodChannel()
        }
        buildSdk(apiKey: apiKey, origin: origin)
    }

}

public struct TikiSdkError : Error {

    var message: String?

    init(message: String?){
        self.message = message
    }
}

File: tiki_sdk_flutter_plugin/android/src/main/kotlin/com/mytiki/tiki_sdk_flutter_plugin/TikiSdkFlutterChannel.kt
package com.mytiki.tiki_sdk_flutter_plugin

import android.content.Context
import io.flutter.embedding.engine.FlutterEngine
import io.flutter.embedding.engine.plugins.FlutterPlugin
import io.flutter.embedding.engine.plugins.FlutterPlugin.FlutterPluginBinding
import io.flutter.plugin.common.BinaryMessenger
import io.flutter.plugin.common.MethodChannel

/**
 * Tiki sdk plugin
 *
 * @property apiKey
 * @property origin
 * @constructor
 *
 * @param context
 */
class TikiSdkFlutterChannel(
    var apiKey: String? = null,
    var origin: String? = null,
    context: Context? = null
) : FlutterPlugin, MethodChannel.MethodCallHandler  {

    var caller: TikiSdkPluginCaller? = null

    private var flutterEngine: FlutterEngine? = null
    private var methodChannel: MethodChannel? = null

    companion object {
        const val channelId = "tiki_sdk_flutter"
    }

    init {
        if (context != null) {
            setupChannel(context)
        }
    }

    private fun setupChannel(context: Context) {
        if (methodChannel == null) {
            if (flutterEngine == null) flutterEngine = FlutterEngine(context)
            methodChannel = MethodChannel(flutterEngine!!.dartExecutor, channelId)
        }
        buildSdk()
    }

    private fun setupChannel(messenger: BinaryMessenger) {
        methodChannel = MethodChannel(messenger, channelId)
        buildSdk()
    }

    private fun buildSdk() {
        caller = TikiSdkPluginCaller(methodChannel!!)
        methodChannel!!.setMethodCallHandler(TikiSdkPluginHandler(caller!!))
        methodChannel!!.invokeMethod(
            "build", mapOf(
                "apiKey" to apiKey,
                "origin" to origin
            )
        )
    }

    private fun teardownChannel() {
        methodChannel!!.setMethodCallHandler(null)
        methodChannel = null
        caller = null
    }

    override fun onMethodCall(@NonNull call: MethodCall, @NonNull result: MethodChannel.Result) {
        val requestId = call.argument<String>("requestId")
        val response = call.argument<String?>("response")
        if (requestId == null) result.error("-1", "missing requestId argument", call.arguments)
        when (call.method) {
            "success" -> {
                caller.completables[requestId]?.complete(response)
            }
            "error" -> {
                caller.completables[requestId]?.completeExceptionally(Exception(response))
            }
            else -> result.notImplemented()
        }
        caller.completables.remove(requestId)
    }

    override fun onAttachedToEngine(binding: FlutterPluginBinding) {
        setupChannel(binding.binaryMessenger)
    }

    override fun onDetachedFromEngine(binding: FlutterPluginBinding) {
        teardownChannel()
    }

}
